lib/internal/querystring.js

```diff
@@ -0,0 +1,165 @@
+  if (typeof str !== 'string') {
+    if (typeof str === 'object')
+      str = String(str);
+    else
+      str += '';
+  }
+  var out = '';
+  var lastPos = 0;
+
+  for (var i = 0; i < str.length; ++i) {
+    var c = str.charCodeAt(i);
+
+    // ASCII
+    if (c < 0x80) {
+      if (noEscape[c] === 1)
+        continue;
+      if (lastPos < i)
+        out += str.slice(lastPos, i);
+      lastPos = i + 1;
+      out += hexTable[c];
+      continue;
+    }
+
+    if (lastPos < i)
+      out += str.slice(lastPos, i);
+
+    // Multi-byte characters ...
+    if (c < 0x800) {
+      lastPos = i + 1;
+      out += hexTable[0xC0 | (c >> 6)] + hexTable[0x80 | (c & 0x3F)];
+      continue;
+    }
+    if (c < 0xD800 || c >= 0xE000) {
+      lastPos = i + 1;
+      out += hexTable[0xE0 | (c >> 12)] +
+             hexTable[0x80 | ((c >> 6) & 0x3F)] +
+             hexTable[0x80 | (c & 0x3F)];
+      continue;
+    }
+    // Surrogate pair
+    ++i;
+    var c2;
+    if (i < str.length)
+      c2 = str.charCodeAt(i) & 0x3FF;
+    else
+      throw new URIError('URI malformed');
+    lastPos = i + 1;
+    c = 0x10000 + (((c & 0x3FF) << 10) | c2);
+    out += hexTable[0xF0 | (c >> 18)] +
+           hexTable[0x80 | ((c >> 12) & 0x3F)] +
+           hexTable[0x80 | ((c >> 6) & 0x3F)] +
+           hexTable[0x80 | (c & 0x3F)];
+  }
+  if (lastPos === 0)
+    return str;
+  if (lastPos < str.length)
+    return out + str.slice(lastPos);
+  return out;
+}
+
+// a safe fast alternative to decodeURIComponent
+function unescapeBuffer(s, decodeSpaces) {
+  var out = Buffer.allocUnsafe(s.length);
+  var state = 0;
+  var n, m, hexchar, c;
+
+  for (var inIndex = 0, outIndex = 0; ; inIndex++) {
+    if (inIndex < s.length) {
+      c = s.charCodeAt(inIndex);
+    } else {
+      if (state > 0) {
+        out[outIndex++] = 37/*%*/;
+        if (state === 2)
+          out[outIndex++] = hexchar;
+      }
+      break;
+    }
+    switch (state) {
+      case 0: // Any character
+        switch (c) {
+          case 37: // '%'
+            n = 0;
+            m = 0;
+            state = 1;
+            break;
+          case 43: // '+'
+            if (decodeSpaces)
+              c = 32; // ' '
+            // falls through
+          default:
+            out[outIndex++] = c;
+            break;
+        }
+        break;
+
+      case 1: // First hex digit
+        hexchar = c;
+        n = unhexTable[c];
+        if (!(n >= 0)) {
+          out[outIndex++] = 37/*%*/;
+          out[outIndex++] = c;
+          state = 0;
+          break;
+        }
+        state = 2;
+        break;
+
+      case 2: // Second hex digit
+        state = 0;
+        m = unhexTable[c];
+        if (!(m >= 0)) {
+          out[outIndex++] = 37/*%*/;
+          out[outIndex++] = hexchar;
+          out[outIndex++] = c;
+          break;
+        }
+        out[outIndex++] = 16 * n + m;
+        break;
+    }
+  }
+
+  // TODO support returning arbitrary buffers.
```

Comment:
I assume it means using a user-provided buffer. However, at the same time, this comment goes all the way back to 57d817290649a53472664cd266d5f756e8e59e95 (v0.3.2), so the applicability of this TODO isn't clear by any means.

Anyone is welcome to initiate a formal deprecation process of these exports in another PR.